src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -377,6 +384,99 @@ public void WriteValue(string value)
+        {
+            int chunkSize = bufferWriter.FreeCapacity;
+
+            for (int i = 0; i < value.Length; i += chunkSize)
+            {
+                int remainingChars = Math.Min(chunkSize, value.Length - i);
+                ReadOnlySpan<char> chunk = value.Slice(i, remainingChars);
+
+                if (NeedsEscaping(chunk) != -1)
+                {
+                    WriteEscapedStringChunk(chunk);
+                }
+                else
+                {
+                    WriteStringChunk(chunk);
+                }
+
+                // Flush the buffer if needed
+                this.FlushIfBufferThresholdReached();
+            }
+        }
+
+        private int NeedsEscaping(ReadOnlySpan<char> value)
+        {
+            if (value.IsEmpty)
+            {
+                return -1;
+            }
+
+            // Iterate through each character in the span
+            for (int i = 0; i < value.Length; i++)
+            {
+                string encoded = this.encoder.Encode(value[i].ToString());
+                
+                // if encoded string is different from original string then we know special
+                // characters exist and need escaping.
+                if (!string.Equals(value[i].ToString(), encoded, StringComparison.Ordinal))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private void WriteEscapedStringChunk(ReadOnlySpan<char> sourceChunk)
+        {
+            char[] valueArray = null;
+            const int maxSize = 0; 
+            const int threshold = 120;
+
+            Span<char> destination = maxSize <= threshold
+                ? stackalloc char[maxSize]
+                :valueArray = ArrayPool<char>.Shared.Rent(maxSize);
+
+            try
+            {
+                OperationStatus status = this.encoder.EncodeUtf8(
+                    MemoryMarshal.Cast<char, byte>(sourceChunk),
+                    MemoryMarshal.Cast<char, byte>(destination), // Use the byte span of the destination
+                    out int bytesConsumed,
+                    out int bytesWritten,
+                    isFinalBlock: true
+                );
+
+                if (status == OperationStatus.Done)
+                {
+                    WriteStringChunk(destination.Slice(0, bytesWritten));
+                }
+                else
+                {
+                    Console.WriteLine($"Encoding failed with status: {status}");
+                }
+            }
+            finally
+            {
+                if (valueArray != null)
+                {
+                    ArrayPool<char>.Shared.Return(destination.ToArray());
+                }
+            }
+        }
+
+        private void WriteStringChunk(ReadOnlySpan<char> chunk)
+        {
+            // Create a buffer to hold UTF-8 encoded bytes
+            byte[] utf8Bytes = new byte[Encoding.UTF8.GetByteCount(chunk)];

```

Comment:
Instead of allocating a new array, we could rent it from the ArrayPool. But since we're going to write these utf8Bytes to the bufferWriter, maybe we should write it to the buffer directly. Here's an example:

```c#
int maxUtf8Length = Encoding.UTF8.GetMaxByteCount(chunk);

// technically this might resize the buffer. But if we use a small chunk size, say 4K, and since we're flushing
// after every chunk, this is unlikely even if we consider the max byte count
Span<byte> output  = bufferWriter.GetSpan(maxUtf8Length);
OperationStatus status = Utf8.FromUtf16(chunk, output, out int charsRead, out int bytesWritten, isFinalBlock: false /* set to true on the last chunk */);
Debug.Assert(status == OperationStatus.Done);
Debug.Assert(charsRead == chunk.Length); // did we read everything in the chunk

// notify the bufferWriter of the write
bufferWriter.Advance(bytesWritten);
```